File: array/ts/plus_one.ts
// 66 加一
// https://leetcode.cn/problems/plus-one/
// 从最低位开始逐位加 1，k 表示当前进位。
// 个位单独处理（9 则写 0 进 1，否则直接 +1），其余位从右向左依次处理：
// 注意必须先把 digits[i] + k 写入结果、再清空 k，否则进位会丢失。
function plusOne(digits: number[]): number[] {
    const n = digits.length;
    const ans: number[] = [];

    let k = 0; // 当前进位

    if (digits[n - 1] === 9) {
        ans.push(0);
        k = 1;
    } else {
        ans.push(digits[n - 1] + 1);
        k = 0;
    }

    for (let i = n - 2; i >= 0; i--) {
        if (digits[i] + k === 10) {
            ans.push(0);
            k = 1;
        } else {
            ans.push(digits[i] + k);
            k = 0;
        }
    }

    if (k === 1) {
        ans.push(1);
    }

    return ans.reverse();
}

File: array/python/plus_one.py
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        n = len(digits)
        ans = []

        k = 0

        if digits[n-1] ==9:
            ans.append(0)
            k =1
        else:
            ans.append(digits[n-1] + 1)
            k =0

        for i in range (n-2,-1,-1):
            if digits[i] + k == 10 :
                ans.append(0)
                k = 1
            else:
                ans.append(digits[i] + k)
                k = 0

        if k == 1:
            ans.append(1)


        ans.reverse()

        return ans
            
        
            